Emby.Server.Implementations/Plugins/PluginManager.cs

```diff
@@ -0,0 +1,694 @@
+            }
+
+            // Update the manifest on disk
+            if (ChangePluginState(plugin, PluginStatus.Disabled))
+            {
+                UpdateSuccessors(plugin);
+            }
+        }
+
+        /// <summary>
+        /// Changes the status of the other versions of the plugin to "Superceded".
+        /// </summary>
+        /// <param name="plugin">The <see cref="LocalPlugin"/> that's master.</param>
+        private void UpdateSuccessors(LocalPlugin plugin)
+        {
+            // This value is memory only - so that the web will show restart required.
+            plugin.Manifest.Status = PluginStatus.RestartRequired;
+
+            // Detect whether there is another version of this plugin that needs disabling.
+            var predecessor = _plugins.OrderByDescending(p => p.Version)
+                .FirstOrDefault(
+                    p => p.Id.Equals(plugin.Id)
+                    && p.IsEnabledAndSupported
+                    && p.Version != plugin.Version);
+
+            if (predecessor == null)
+            {
+                return;
+            }
+
+            if (predecessor.Manifest.Status == PluginStatus.Active && !ChangePluginState(predecessor, PluginStatus.Superceded))
+            {
+                _logger.LogError("Unable to disable version {Version} of {Name}", predecessor.Version, predecessor.Name);
+            }
+        }
+
+        /// <summary>
+        /// Disable the plugin.
+        /// </summary>
+        /// <param name="assembly">The <see cref="Assembly"/> of the plug to disable.</param>
+        public void FailPlugin(Assembly assembly)
+        {
+            // Only save if disabled.
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            var plugin = _plugins.Where(
+                    p => assembly.Equals(p.Assembly)
+                    || string.Equals(assembly.Location, assembly.Location, StringComparison.OrdinalIgnoreCase))
+                .FirstOrDefault();
+            if (plugin == null)
+            {
+                // A plugin's assembly didn't cause this issue, so ignore it.
+                return;
+            }
+
+            ChangePluginState(plugin, PluginStatus.Malfunction);
+        }
+
+        /// <summary>
+        /// Saves the manifest back to disk.
+        /// </summary>
+        /// <param name="manifest">The <see cref="PluginManifest"/> to save.</param>
+        /// <param name="path">The path where to save the manifest.</param>
+        /// <returns>True if successful.</returns>
+        public bool SaveManifest(PluginManifest manifest, string path)
+        {
+            if (manifest == null)
+            {
+                return false;
+            }
+
+            try
+            {
+                var data = JsonSerializer.Serialize(manifest, _jsonOptions);
+                File.WriteAllText(Path.Combine(path, "meta.json"), data, Encoding.UTF8);
```

Comment:
We should check if the metadata changed before writing the file.

Reply:
This is now checked in all the calling functions.